Cython/Compiler/Nodes.py

```diff
@@ -1860,8 +1861,13 @@ class FuncDefNode(StatNode, BlockNode):
     #       by AnalyseDeclarationsTransform, so it can replace CFuncDefNodes
     #       with fused argument types with a FusedCFuncDefNode
 
+    class NeedsClosure(enum.IntEnum):
+        NO_CLOSURE = 0
+        GENERATOR_ONLY = 1
+        FULL_CLOSURE = 2
```

Comment:
I want to distinguish between the types of closure. Things that are purely generators don't need locking because generators have an `is_running` check.

Reply:
I think this is a good distinction.